### PR TITLE
Skeleton of GRPO

### PR DESCRIPTION
**Pull request by joecummings**
### What this PR DOES do:

**Core GRPO Implementation:**

*   Complete **GRPO** training system for RL fine-tuning
*   Integration with **vLLM policy actors** for text generation during rollouts
*   **Multi-reward system** supporting math correctness and thinking tag rewards
*   **Replay buffer** with configurable batch sizes and policy version tracking
*   **Async training loops** with concurrent rollout generation and policy training

**Service Infrastructure:**

*   **Service-based architecture** using Monarch actors for distributed components
*   **Policy actors** with configurable sampling, device assignment, and tensor parallelism
*   **Trainer actors** with GRPO loss computation, KL regularization, and gradient clipping
*   **Reference model actors** for computing baseline log probabilities
*   **Advantage computation** using reward-to-go with normalization
*   Dataset integration with **GSM8K** math problems

### What this PR will NOT cover:

**Missing Features:**

*   **Weight synchronization** between trainer and policy (commented out: `# await trainer.update_weights(policy)`) @pradeepfn @pbontrager 
*   **Policy versioning** system (hardcoded to version 0) @joecummings @Jack-Khuu 
*   **Logging/monitoring** infrastructure (No wandb / tensorboard integration) @calvinpelletier @DNXie 
*   **Multi-turn conversations** or complex dialog handling 
*   **Production-ready error handling** and recovery mechanisms @allenwang28 

**Incomplete Work:**

*   **TODO:** Move policy processing initialization into setup (line 453 in [`**main.py**`](command:code-compose.open?%5B%22%2Fhome%2Fjrcummings%2Fprojects%2Fjoe-forge%2Fapps%2Fgrpo%2Fmain.py%22%2Cnull%5D "apps/grpo/main.py")) @Jack-Khuu 
*   **Manual device assignment** instead of automatic GPU allocation @pbontrager 
*   **Hardcoded hyperparameters** throughout the system

**Technical Limitations:**

*   Single model support (Qwen3-1.7B)
*   Fixed reward functions (no pluggable reward interface) @DNXie 
*   No distributed training across multiple machines 

This is a **working prototype** of GRPO with the core RL loop functional but requiring additional work for production deployment, proper weight updates, and comprehensive logging.

### CODE REVIEW

File: apps/grpo/main.py [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+                wandb.log({"rollout_count": rollout_count, "avg_reward": avg_reward})
+                # print(
+                #     f"Generated {rollout_count} rollouts w/ average reward {avg_reward}"
+                # )
+
+    async def continuous_training():
+        training_step = 0
+        while True:
+            batch = await replay_buffer.sample.choose(curr_policy_version=0)
+            if batch is None:
```
Comment (pbontrager) [inline apps/grpo/main.py:506]: I feel this should just be inside of the buffer.sample method. Then you'll just await until there's enough data for a sample.

File: apps/grpo/main.py [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+    reward_actor = await spawn_service(
+        default_service_cfg,
+        RewardActor,
+        scoring_functions=[math_scoring_function, thinking_scoring_function],
+    )
+
+    print("All services initialized successfully!")
+
+    # Intialize logging
+    wandb.init(project="forge", name="grpo_test")
```
Comment (ebsmothers) [inline apps/grpo/main.py]: Can merge with metric logging PR (#59) to get rid of this?

File: apps/grpo/main.py [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+        sequence_log_probs = compute_sequence_logprobs(
+            self.model, input_ids, attention_mask, requires_grad=False
+        )
+
+        return (
+            sequence_log_probs.squeeze()
+        )  # Remove batch dimension for single response
+
+
+class DatasetActor(ForgeActor):
```
Comment (ebsmothers) [inline apps/grpo/main.py:359]: Longer-term what is our plan here? (I.e. is our HfIterableDataset compatible with how we're setting things up here?)

File: apps/grpo/main.py [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+    def _setup_dataset(self, *args, **kwargs):
+        def gsm8k_to_messages(sample):
+            question = sample["question"]
+            full_answer: str = sample["answer"]
+            answer = full_answer.split("#### ")[1]
+            return {"question": question, "answer": answer}
+
+        ds = load_dataset(*args, **kwargs)
+        ds = ds.map(gsm8k_to_messages)
+        ds = ds.shuffle()
```
Comment (ebsmothers) [inline apps/grpo/main.py:375]: I notice no `split_dataset_by_node`. Is this by design?

File: apps/grpo/main.py [hunk 5]
```diff
@@ -0,0 +1,581 @@
...
+            self.device = device
+
+        # Initialize model and tokenizer
+        self.model = AutoModelForCausalLM.from_pretrained(
+            model_name,
+            torch_dtype=torch.float16,
+            trust_remote_code=True,
+        ).to(self.device)
+
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name)
```
Comment (ebsmothers) [inline apps/grpo/main.py]: Why do we want tokenizer in the reference model?

File: apps/grpo/main.py [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+            version = 0  # await policy.get_current_version.choose()
+            episode = Episode(
+                episode_id=rollout_count,
+                prompt=prompt,
+                target=target,
+                policy_version=version,
+            )
+
+            # async with policy.session(version=version):
+            #     actions: list[CompletionOutput] = await policy.generate(prompt)
```
Comment (ebsmothers) [inline apps/grpo/main.py]: What's preventing us from doing this?

File: apps/grpo/main.py [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+                group.advantage = advantage
+
+            await replay_buffer.add.choose(episode)
+
+            rollout_count += 1
+            if rollout_count % 10 == 0:
+                avg_reward = sum(group.reward for group in episode.groups) / len(
+                    episode.groups
+                )
+                wandb.log({"rollout_count": rollout_count, "avg_reward": avg_reward})
```
Comment (ebsmothers) [inline apps/grpo/main.py]: Will this eventually be its own actor?

File: apps/grpo/main.py [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+        self.model_name = model_name
+
+        # Set device
+        if device is None:
+            self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        else:
+            self.device = device
+
+        # Initialize model and tokenizer
+        self.model = AutoModelForCausalLM.from_pretrained(
```
Comment (ebsmothers) [inline apps/grpo/main.py:96]: Have you tried hooking up to titan model definitions? Or is that out of scope for this PR

File: apps/grpo/main.py [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+async def main():
+    """Main GRPO training loop with rollout and training processes."""
+    group_size = 5
+    model = "Qwen/Qwen3-1.7B"
+    # all_gpus = torch.cuda.device_count()
+
+    # ---- Setup services ---- #
+    default_service_cfg = ServiceConfig(
+        procs_per_replica=1,
+        num_replicas=1,
```
Comment (ebsmothers) [inline apps/grpo/main.py:394]: Can we vary any of these yet? E.g. I see in some of the actors that you've hardcoded device info etc. Would be good to take that out asap

File: apps/grpo/main.py [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+class ComputeAdvantages(ForgeActor):
+    """Compute advantages for GRPO using reward signals."""
+
+    def __init__(self, gamma: float = 0.99, lambda_: float = 0.95):
+        super().__init__()
+        self.gamma = gamma  # Discount factor
+        self.lambda_ = lambda_  # GAE lambda parameter
+
+    @endpoint
+    async def __call__(self, groups: list[Group]) -> list[float]:
```
Comment (ebsmothers) [inline apps/grpo/main.py:285]: Is this based off of some specific reference implementation?

File: src/forge/data/replay_buffer.py [hunk 11]
```diff
@@ -9,60 +9,55 @@
...
         self.batch_size = batch_size
         self.max_policy_age = max_policy_age
         if seed is not None:
             random.seed(seed)
         self.sampler = random.sample
 
     @endpoint
-    async def add(self, trajectory: Trajectory) -> None:
-        self.buffer.append(trajectory)
+    async def add(self, episode) -> None:
```
Comment (ebsmothers) [inline src/forge/data/replay_buffer.py]: So we are moving from Trajectories to Episodes in the replay buffer? Isn't this overly-tailored to GRPO?

File: apps/grpo/main.py (hunk 1 above)
Comment (joecummings) [reply apps/grpo/main.py:506]: Hmm I disagree. The contract here is just that the buffer will return a sample here when it computationally can, not when there's something inside. 

In addition, this would push possible errors a layer down if the replay buffer isn't getting filled for some reason. I'd rather have this be exposed logic to the user.

File: apps/grpo/main.py (hunk 2 above)
Comment (joecummings) [reply apps/grpo/main.py]: No, I'd rather keep things insanely minimal to start.

File: apps/grpo/main.py (hunk 3 above)
Comment (joecummings) [reply apps/grpo/main.py:359]: I think we need to re-build up our HfIterableDataset based on the patterns we observe when testing out this training loop. Right now, it's not super intuitive to use and it's also not clear what we actually *need* from there. 

My guess is that we'll keep 75% of the current implementation, but I don't want to assume before we actually start testing.

File: apps/grpo/main.py (hunk 4 above)
Comment (joecummings) [reply apps/grpo/main.py:375]: No, just missing it. Will add.

File: apps/grpo/main.py (hunk 5 above)
Comment (joecummings) [reply apps/grpo/main.py]: Shouldn't be needed actually b/c I believe we can get what we want from the tokens returned from vllm.

File: apps/grpo/main.py (hunk 6 above)
Comment (joecummings) [reply apps/grpo/main.py]: #70

File: apps/grpo/main.py (hunk 7 above)
Comment (joecummings) [reply apps/grpo/main.py]: I'm not sure it needs to be unless we're making so many requests from all over to log things and we want it going through a central source. 

Otherwise, we can just treat the logger as a normal component.

File: apps/grpo/main.py (hunk 8 above)
Comment (joecummings) [reply apps/grpo/main.py:96]: Waaaay out of scope.

File: apps/grpo/main.py (hunk 9 above)
Comment (joecummings) [reply apps/grpo/main.py:394]: I want to get everything working first and then update these.

File: apps/grpo/main.py (hunk 10 above)
Comment (joecummings) [reply apps/grpo/main.py:285]: My brain, the old impl of GRPO in torchtune, and good ol devmate.

File: src/forge/data/replay_buffer.py (hunk 11 above)
Comment (joecummings) [reply src/forge/data/replay_buffer.py]: Everything in here is overly tailored to GRPO. I want to base our design decisions from real life, otherwise we can talk all day. This is definitely not the final state of our APIs

File: src/forge/actors/policy.py [hunk 12]
```diff
@@ -359,6 +365,7 @@ def setup_worker(self):
             "distributed_init_method": distributed_init_method,
             "is_driver_worker": is_driver_worker,
         }
+        print(all_kwargs)
```
Comment (DNXie) [inline src/forge/actors/policy.py]: Is this line for debugging? Should we remove this?

File: apps/grpo/main.py (hunk 10 above)
Comment (ebsmothers) [reply apps/grpo/main.py:285]: I'm not sure I trust that consensus..

File: apps/grpo/main.py (hunk 5 above)
Comment (allenwang28) [reply apps/grpo/main.py]: The ref logits should be from the base model, so we shouldn't be able to get it from vLLM? Unless we spin up vLLM actors specifically for getting the logits (I'm not opposed to this either)

cc @pbontrager

File: apps/grpo/main.py (hunk 7 above)
Comment (allenwang28) [reply apps/grpo/main.py]: would wandb complain about having multiple "sessions" created from the same job? If we can minimize the messages being passed in Monarch that generally seems like a good idea

File: apps/grpo/main.py (hunk 1 above)
Comment (pbontrager) [reply apps/grpo/main.py:506]: When does sample return None? Is it when number of usable rollouts < batch_size? I feel like this is going to surprise some people who try to use this and start running into Nones in their batch.

File: apps/grpo/main.py (hunk 2 above)
Comment (pbontrager) [reply apps/grpo/main.py]: Things like that can be changed by others after this code lands

File: apps/grpo/main.py (hunk 5 above)
Comment (pbontrager) [reply apps/grpo/main.py]: These are the tokens not the logits, that should be fine most of the time.

File: apps/grpo/main.py (hunk 7 above)
Comment (pbontrager) [reply apps/grpo/main.py]: I think we want it to be it's own actor so logs are called from within actors too. It's a lot of unnecessary data passing otherwise, especially if people log artifacts

File: apps/grpo/main.py (hunk 10 above)
Comment (pbontrager) [reply apps/grpo/main.py:285]: We're probably not going to want a Group type, but we put the type work on the back burner for now

File: src/forge/data/replay_buffer.py (hunk 11 above)
Comment (pbontrager) [reply src/forge/data/replay_buffer.py]: Episode isn't GRPO specific, it's basically another word for Trajectory.

File: apps/grpo/main.py [hunk 13]
```diff
@@ -0,0 +1,581 @@
...
+    async def continuous_rollouts():
+        rollout_count = 0
+        while True:
+            sample = await dataloader.__next__.choose()
+            if sample is None:
+                print("Dataloader is empty, exiting continuous rollout")
+                return
+            print(sample)
+            prompt, target = sample
+            version = 0  # await policy.get_current_version.choose()
```
Comment (DNXie) [inline apps/grpo/main.py:466]: version is never updated. Is this correct?

File: apps/grpo/main.py (hunk 10 above)
Comment (allenwang28) [reply apps/grpo/main.py:285]: we should spin up a track fully dedicated to ensuring numerical correctness (which I think is the basis of @ebsmothers comment), but imo the right timing for this is after we have the working prototype

File: apps/grpo/main.py (hunk 7 above)
Comment (allenwang28) [reply apps/grpo/main.py]: hmm, I'm not sure if I understand that comment @pbontrager 

if the metrics logger is an actor, then all actors would need to pass messages over Monarch to the metrics actor right? If actors can just write directly to wandb that would minimize data passing?

File: apps/grpo/main.py (hunk 1 above)
Comment (joecummings) [reply apps/grpo/main.py:506]: Yeah if the replay buffer has nothing in it

File: apps/grpo/main.py (hunk 13 above)
Comment (joecummings) [reply apps/grpo/main.py:466]: Will fix later!

File: apps/grpo/main.py [hunk 14]
```diff
@@ -0,0 +1,525 @@
+import asyncio
```
Comment (allenwang28) [inline apps/grpo/main.py:7]: nit: add the copyright string thing
```suggestion
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

"""To run:

python -m apps.grpo.main
"""
import asyncio
```

File: apps/grpo/main.py [hunk 15]
```diff
@@ -0,0 +1,525 @@
...
+
+class Episode:
+    """Episode container for GRPO rollouts."""
+
+    def __init__(self, episode_id: int, prompt: str, target: str, policy_version: int):
+        self.episode_id = episode_id
+        self.prompt = prompt
+        self.target = target
+        self.policy_version = policy_version
+        self.groups: list[Group] = []
```
Comment (allenwang28) [inline apps/grpo/main.py:68]: naming nit, but isn't what we're calling `Group` here actually an individual output, and what we're calling `self.groups` the actual `Group`?

File: apps/grpo/main.py [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+
+        for episode in batch:
+            groups = episode.groups
+
+            # Collect all response texts and corresponding data
+            response_texts = []
+            ref_logprobs_list = []
+            advantages_list = []
+
+            for group in groups:
```
Comment (allenwang28) [inline apps/grpo/main.py:127]: no action needed now, but imo it would be nifty for the episode structure itself to handle things like this

```
for episode in batch:
    tokenized = self.tokenizer(
        episode.as_response_group(),
        ...
    )
```

File: apps/grpo/main.py (hunk 1 above)
Comment (allenwang28) [reply apps/grpo/main.py:506]: I think exposing logic to the user makes sense, my only nit is that I might want it to have more of a queue-like semantic? Like we set a timeout, it raises an error if we hit the timeout etc.

File: src/forge/actors/policy.py [hunk 17]
```diff
@@ -106,6 +103,11 @@ class Policy(PolicyInterface):
     @endpoint
     async def setup(self):
         # Set up policy_worker
+        self.available_devices = (
```
Comment (pbontrager) [inline src/forge/actors/policy.py:107]: This should really come from the script and not the config. Ideally the script makes the master set of ranks and then passes them to the actors.

File: apps/grpo/main.py (hunk 15 above)
Comment (joecummings) [reply apps/grpo/main.py:68]: Fair lol

File: apps/grpo/main.py (hunk 1 above)
Comment (joecummings) [reply apps/grpo/main.py:506]: Does `choose` have a built-in timeout feature?

Comment (allenwang28) [reply apps/grpo/main.py:506]: it doesn't, but the buffer itself can return an exception which will be propagated through choose

but the current service implementation will then mark the replica as unhealthy so we shouldn't add this yet
